### PR TITLE
Feat/graph chaoda compile

### PR DESCRIPTION
**Pull request by djperrone**
Changed function signatures and type aliases in graph and chaoda modules to match current cluster struct. Changed most function bodies to be todo()! because they need refactoring to pass pre-commit checks. Added documentation to both modules. Added crates smartcore, ndarray, automl to cargo.toml. Added .idea to gitignore.

### CODE REVIEW

File: abd-clam/Cargo.toml [hunk 1]
```diff
@@ -43,15 +43,17 @@ priority-queue = "1.3.2"
 # TODO: Break CHAODA out into an optional feature
 # automl = { version = "0.3.0", path = "../rust-automl" }
 # automl = { git = "https://github.com/nishaq503/rust-automl", branch = "updates/dependencies" }
+automl = "0.2.7"
```
Comment (nishaq503) [inline abd-clam/Cargo.toml:46]: Add a TODO here for us to switch from `automl` to `smartcore`

File: abd-clam/Cargo.toml [hunk 2]
```diff
@@ -43,15 +43,17 @@ priority-queue = "1.3.2"
...
 # Only used for normalizing cluster ratios for CHAODA
 libm = "0.2.7"
 statistical = "1.0.0"
 
 # Only used for sharded datasets
 # TODO: Break sharded datasets out into an optional feature
-# ndarray = { version = "0.15.6", features = ["rayon"] }
-# ndarray-npy = "0.8.1"
+ndarray = { version = "0.15.6", features = ["rayon"] }
+ndarray-npy = "0.8.1"
```
Comment (nishaq503) [inline abd-clam/Cargo.toml:54]: Remove the `ndarray-npy` dependency and any of its uses in code. We will replace IO of arrays with an in-house solution

File: abd-clam/src/core/cluster/_cluster.rs [hunk 3]
```diff
@@ -15,8 +15,8 @@ use std::fmt::Write;
 use crate::{utils, Dataset, Instance, PartitionCriteria, PartitionCriterion};
 
 /// Ratios are used for anomaly detection and related applications.
-pub type Ratios = [f64; 6];
-
+// pub type Ratios = [f64; 6];
+use crate::core::cluster::Ratios;
```
Comment (nishaq503) [inline abd-clam/src/core/cluster/_cluster.rs:19]: Why did you move where this is defined? I prefer this here but if there's a good reason to move it, that's fine.

File: abd-clam/src/core/graph/criteria.rs [hunk 4]
```diff
@@ -1,4 +1,4 @@
 //! Criteria used for selecting `Cluster`s for `Graph`s.
 
 /// A `Box`ed function that assigns a score for a given `Cluster`.
-pub type MetaMLScorer = fn(crate::cluster::Ratios) -> f64;
+pub type MetaMLScorer = Box<fn(crate::core::cluster::Ratios) -> f64>;
```
Comment (nishaq503) [inline abd-clam/src/core/graph/criteria.rs:4]: Why box the function?

File: abd-clam/src/core/graph/_graph.rs [hunk 5]
```diff
@@ -4,53 +4,64 @@ use std::collections::{HashMap, HashSet};
...
-impl<'a, T: Number, U: Number, D: Dataset<T, U>> Edge<'a, T, U, D> {
+impl<'a, U: Number> Edge<'a, U> {
     /// Creates a new `Edge` from the given `Cluster`s and the distance between
     /// them.
     ///
     /// It is upon the user to verify that the two `Cluster`s are close enough
     /// to have an edge between them.
-    #[allow(dead_code)]
-    pub fn new(left: &'a Cluster<T, U, D>, right: &'a Cluster<T, U, D>, distance: U) -> Self {
+    #[allow(unused_mut, unused_variables, dead_code)]
```
Comment (nishaq503) [inline abd-clam/src/core/graph/_graph.rs:63]: Remove this macro from all code. Make the `Graph` and `Edge` structs public in `lib.rs`

File: abd-clam/src/core/graph/_graph.rs [hunk 6]
```diff
@@ -142,60 +164,66 @@ impl<'a, T: Number, U: Number, D: Dataset<T, U>> Graph<'a, T, U, D> {
...
-            distance_matrix: None,
-            adjacency_matrix: None,
-            frontier_sizes: None,
-        }
-    }
-
-    #[allow(dead_code)]
+    #[allow(unused_mut, unused_variables, dead_code)]
+    #[allow(clippy::needless_pass_by_value)]
+    pub fn new(_clusters: ClusterSet<'a, U>, _edges: EdgeSet<'a, U>) -> Self {
```
Comment (nishaq503) [inline abd-clam/src/core/graph/_graph.rs:169]: Change this signature to use references. Then remove the allow-clippy... macro

File: abd-clam/src/core/graph/_graph.rs [hunk 7]
```diff
@@ -142,60 +164,66 @@ impl<'a, T: Number, U: Number, D: Dataset<T, U>> Graph<'a, T, U, D> {
...
+        //     max_depth,
+        //     distance_matrix: None,
+        //     adjacency_matrix: None,
+        //     frontier_sizes: None,
+        // }
+    }
+
+    #[allow(clippy::missing_docs_in_private_items)]
+    #[allow(unused_mut, unused_variables, dead_code)]
     fn compute_distance_matrix(&self) -> Vec<Vec<U>> {
```
Comment (nishaq503) [inline abd-clam/src/core/graph/_graph.rs:210]: Add docs here and remove the missing docs macro. This function should create a square, symmetric matrix of edge lengths, leaving zero distances for when there is no edge

File: abd-clam/src/core/graph/_graph.rs [hunk 8]
```diff
@@ -218,200 +246,284 @@ impl<'a, T: Number, U: Number, D: Dataset<T, U>> Graph<'a, T, U, D> {
...
+        //     adjacency_matrix: self.adjacency_matrix.clone(),
+        //     frontier_sizes,
+        // }
+    }
+
+    /// Finds and returns connected component clusters within the graph.
+    ///
+    /// This method identifies clusters in the graph that are part of separate connected components
+    /// and returns a vector of sets where each set contains clusters from one connected component.
     #[allow(clippy::manual_retain)]
```
Comment (nishaq503) [inline abd-clam/src/core/graph/_graph.rs:282]: This macro is no longer needed because the code is commented out.

File: abd-clam/src/core/graph/_graph.rs [hunk 9]
```diff
@@ -218,200 +246,284 @@ impl<'a, T: Number, U: Number, D: Dataset<T, U>> Graph<'a, T, U, D> {
...
             Ok(())
         } else {
             Err(format!("Cluster {c} is not in this graph."))
         }
     }
 
-    #[allow(dead_code)]
-    pub fn unchecked_vertex_degree(&'a self, c: &Cluster<T, U, D>) -> usize {
+    /// Returns the vertex degree (number of neighbors) of a given cluster.
+    /// Panics:
```
Comment (nishaq503) [inline abd-clam/src/core/graph/_graph.rs:414]: There should be an empty line before the `Panics:` docs start. This should look like:

```rust
/// A one-line description of the function.
///
/// A longer description
/// of the function.
///
/// # Arguments
///
/// * `arg_1`: The first argument.
/// * `arg_2`: The second argument.
///
/// # Returns:
///
/// If there is any return value.
///
/// # Errors
///
/// If the return is a `Result` type
///
/// # Panics
///
/// * First panic condition.
/// * Second panic condition.
fn fancty_function(...) -> ... { ... }
```

**Review by nishaq503 — CHANGES_REQUESTED** [review]
We need a few changes in code:

- Remove the `#[allow(unused_mut, unused_variables, dead_code)]` macro (and all similar macros from the code.
- Doc-string formatting needs to be better. I have left a template in a comment on the `_graph.rs` file.

There are some other things I want to see changed. After you've changed these first set of things, let me know. I will change a few things myself and open a PR to your branch,

File: abd-clam/src/core/cluster/_cluster.rs (hunk 3 above)
Comment (djperrone) [reply abd-clam/src/core/cluster/_cluster.rs:19]: I had trouble importing the Ratios type into other files. In pretrained_models for example, I could only get the 'use' statement to work when the type alias was in the mod file.

File: abd-clam/src/core/graph/criteria.rs (hunk 4 above)
Comment (djperrone) [reply abd-clam/src/core/graph/criteria.rs:4]: The following block of code boxes the score functions and stores them in a vec but the type of MetaMLScorer is not boxed. I could change the type of the vec to be `Vec<(&'a str, Box<crate::core::graph::MetaMLScorer>)>` instead if you prefer

```
pub fn get_meta_ml_scorers<'a>() -> Vec<(&'a str, crate::core::graph::MetaMLScorer)> {
    vec![
        ("lr_manhattan_sc", Box::new(lr_manhattan_sc)),
```

File: abd-clam/Cargo.toml (hunk 1 above)
Comment (djperrone) [reply abd-clam/Cargo.toml]: done

File: abd-clam/src/core/graph/_graph.rs (hunk 5 above)
Comment (djperrone) [reply abd-clam/src/core/graph/_graph.rs]: Done. I did need to allow dead code for about a dozen fields because of clippy complaining

File: abd-clam/src/core/graph/_graph.rs (hunk 9 above)
Comment (djperrone) [reply abd-clam/src/core/graph/_graph.rs]: I refactored the documentation in graph.rs to match that template. Still need to work on the other files.

File: abd-clam/src/core/cluster/_cluster.rs (hunk 3 above)
Comment (nishaq503) [reply abd-clam/src/core/cluster/_cluster.rs:19]: In that case, remove the commented out lines of code

File: abd-clam/src/core/graph/criteria.rs (hunk 4 above)
Comment (nishaq503) [reply abd-clam/src/core/graph/criteria.rs]: It should be doable without using box but I don't have a preference either way. I was just curious.

File: abd-clam/src/core/graph/_graph.rs (hunk 5 above)
Comment (nishaq503) [reply abd-clam/src/core/graph/_graph.rs]: Ok. In general, let's limit the use of these macros as much as possible.

File: abd-clam/src/core/graph/criteria.rs (hunk 4 above)
Comment (djperrone) [reply abd-clam/src/core/graph/criteria.rs]: Without boxing I get an error. The existing comment above the type alias refers to it as a boxed function as well. If you don't have a preference I'll just leave it as is in the commit

File: abd-clam/src/core/graph/_graph.rs (hunk 5 above)
Comment (djperrone) [reply abd-clam/src/core/graph/_graph.rs]: When I was first trying to get through pre-commit checks clippy wanted most of those macros used. I deleted all of them and then went back and added any macros in where clippy complained. This time it was only about a dozen.

**Review by nishaq503 — APPROVED** [review]
LGTM!
